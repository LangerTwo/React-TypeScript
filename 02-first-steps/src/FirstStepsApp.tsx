import { ItemCounter } from "./shopping-cart/ItemCounter";

interface ItemInCart {
    productName: string;
    quantity: number;
};

const itemsInCart: ItemInCart[] = [
    { productName: 'Nintendo Switch 2', quantity: 1 },
    { productName: 'Pro Controller', quantity: 2 },
    { productName: 'Super Smash', quantity: 3 },
];

export function FirstStepsApp() {

    return (
        <>
            <h1>Carriyo de Compras</h1>

            {
                itemsInCart.map(({ productName, quantity }) => (
                    <ItemCounter key={productName} name={productName} quantity={quantity} />
                ))
            }

            {/* <ItemCounter name="Nintendo Switch 2" quantity={15}/>
            <ItemCounter name="Pro Controller" quantity={20}/>
            <ItemCounter name="Super Smash" quantity={25}/>
            <ItemCounter name="Super Mario" quantity={30}/> */}
        </>
    );
};