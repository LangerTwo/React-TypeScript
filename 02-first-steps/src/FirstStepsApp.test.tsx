import { render, screen } from "@testing-library/react";
import { afterEach, describe, expect, test, vi } from "vitest";
import { FirstStepsApp } from "./FirstStepsApp";
// import { ItemCounter } from "./shopping-cart/ItemCounter";

const mockItemCounter = vi.fn( (_props: unknown) => {
    return (
        <div data-testid="ItemCounter" />
    )
});

vi.mock("./shopping-cart/ItemCounter", () => ({
    ItemCounter: (props: unknown) => mockItemCounter(props)
}));

// vi.mock("./shopping-cart/ItemCounter", () => ({
//     ItemCounter: (props: unknown) => 
//         <div 
//             data-testid="ItemCounter"
//             name={props.name}
//             quantity={props.quantity}
//         />
// }));

describe("FirstStepsApp", () => {
    afterEach(() => {
        vi.clearAllMocks();
    });

    test("should match the snapshot", () => {
        const { container } = render(<FirstStepsApp />);

        expect(container).toMatchSnapshot();
    });

    test("should render the correct number of ItemCounter components", () => {
        render(<FirstStepsApp />);

        const itemCountersEl = screen.getAllByTestId("ItemCounter");

        expect(itemCountersEl.length).toBe(3);
        // screen.debug();
    });

    test("should render ItemCounter with the correct props", () => {
        render(<FirstStepsApp />);

        expect( mockItemCounter ).toHaveBeenCalledTimes(3);
        expect( mockItemCounter ).toHaveBeenCalledWith({
            name: 'Nintendo Switch 2',
            quantity: 1
        });
        expect( mockItemCounter ).toHaveBeenCalledWith({
            name: 'Pro Controller',
            quantity: 2
        });
        expect( mockItemCounter ).toHaveBeenCalledWith({
            name: 'Super Smash',
            quantity: 3
        });
    });
});