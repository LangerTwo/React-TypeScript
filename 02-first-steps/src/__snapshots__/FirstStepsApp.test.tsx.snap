// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FirstStepsApp > should match the snapshot 1`] = `
<div>
  <h1>
    Carriyo de Compras
  </h1>
  <div
    data-testid="ItemCounter"
  />
  <div
    data-testid="ItemCounter"
  />
  <div
    data-testid="ItemCounter"
  />
</div>
`;
